keps/sig-cloud-provider/20191004-out-of-tree-credential-providers.md

```diff
@@ -0,0 +1,216 @@
+
+A RegistryCredentialProvider configuration API type (similar to [clientauthentication](https://github.com/kubernetes/kubernetes/tree/0273d43ae9486e9d0be292c01de2dd4143522b86/staging/src/k8s.io/client-go/pkg/apis/clientauthentication/v1beta1)) will be added to Kubernetes:
+
+```go
+// RegistryCredentialProvider is used by the kubelet container runtime to match image URLs with
+// exec-based credential provider plugins that provide container registry credentials.
+type RegistryCredentialProvider struct {
+    metav1.TypeMeta `json:",inline"`
+
+    // ImageURLMatchers is a list of strings used to match against the image URL
+    // to determine which images to provide credentials for.  If one of the
+    // string matches the image URL, then the RegistryCredentialProvider will be
+    // used by kubelet to provide credentials for the image pull.
+    //
+    // Each ImageURLMatcher string is a URL pattern which can optionally contain
+    // a port and a path.  Globs can be used in the hostname (but not the port or
+    // the path).
+    //
+    // An image URL matches when it has the same number of parts as the
+    // ImageURLMatcher string, and each part matches.  Additionally the path of
+    // ImageURLMatcher must be a prefix of the target URL. If the ImageURLMatch
+    // contains a port, then the port must match as well.
+    ImageURLMatch []string `json:"imageURLMatch,omitempty"`
+
+    // Exec specifies a custom exec-based plugin.  This type is defined in
+    // https://github.com/kubernetes/client-go/blob/62f256057db7571c5ed1aba47eea291f72dd557a/tools/clientcmd/api/types.go#L184
+    Exec clientcmd.ExecConfig
+}
+```
+
+The RegistryCredentialProvider will be encoded in YAML and located in a file on disk.  The directory where credential provider configuration is found will be passed to kubelet via a new configuration option `RegistryCredentialProviderDir`.
+
+We will create new types `RegistryCredentialPluginRequest` and `RegistryCredentialPluginResponse` which will define the interface between the plugin and the kubelet runtime.  After the kubelet matches the image URL to a RegistryCredentialProvider, the kubelet will exec the plugin binary, and pass the JSON encoded request to the plugin via stdin.  This includes the image URL that is to be pulled.  The plugin will report back the response, which includes the credentials that kubelet needs to pull the image.
+
+In the in-tree implementation, the docker keyring, which has N credential providers, returns an `[]AuthConfig` on a Lookup(image string) call.  This struct will be populated by the plugin rather than the in-tree provider.
+
+```go
+// RegistryCredentialPluginRequest is passed to the plugin via stdin, and includes the image that will be pulled by kubelet.
+type RegistryCredentialPluginRequest struct {
+    // Image is used when passed to registry credential providers as part of an
+    // image pull
+    Image string `json:"image"`
```

Comment:
is this intended to match the `pod.spec.containers[*].image` string, or is this a transformed/resolved value?

Reply:
@smarterclayton @mikedanese - was it a goal to have service account name/namespace info available here for dynamic image pull credentials, or am I misremembering?

Reply:
> was it a goal to have service account name/namespace info available here for dynamic image pull credentials, or am I misremembering?

I'm also ok with deferring that, if that information is not available to global credential providers today, and moving forward with just the image would unblock dropping cloud provider dependencies

Reply:
Re https://github.com/kubernetes/kubernetes/issues/68810: if we choose to implement that as a image credential provider, a service account token would be ideal. I was envisioning that this flow would be requested via the API though (e.g. something like image pull secret). I don't think it's feasible to expect something like that to be enabled on a dime via node level configuration.

Reply:
If the CRI implementation was able to validate (and make authorization decisions based on) a bound SA token, then we could just have the kubelet pass such a token and skip the exec logic all together?

The CRI implementation would need to make kube API calls or we would have to encode more claims within the token to allow for authz checks.